rfcs/text/0428-extending-and-duplicating-types.md

```diff
@@ -0,0 +1,991 @@
+
+When applied simultaneously, these constraints are _incompatible_ and result in an unsatisfiable type. Let us try and create an instance that satisfies it. First we try and create one with a string `name`:
+
+```json5
+{
+  name: "John Doe", // ERROR: expected string array, got string
+  ... // superpower, age, occupation
+}
+```
+
+And then we instead try and create one with an array of `name`s:
+
+```json5
+{
+  name: ["John Doe"], // ERROR: expected string, got string array
+  ... // superpower, age, occupation
+}
+```
+
+### An uncomfortable compromise
+
+As alluded to above, checking compatibility of any two given schemas is quite possibly an NP-Hard problem. Even if it is generally solvable, it is likely a task that is complicated enough that it would be an exceptionally high barrier to entry for potential implementers to overcome.
+
+While various implementations likely will want to implement checking for the simpler cases (disallowing a property from being two disjoint types for example), we opt not to encode such restrictions within the type system, and **implementations should not assume that an arbitrary type is satisfiable**.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+- [JSON Schema composition](https://json-schema.org/understanding-json-schema/reference/combining.html) and [`unevaluatedProperties`](https://json-schema.org/understanding-json-schema/reference/object.html#unevaluated-properties)
+- [Programming languages subtyping](https://en.wikipedia.org/wiki/Subtyping)
+- [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- We haven't specified how projecting/selecting properties of a supertype from a subtype instance is possible. It is an open question how we actually pick out the exact properties of a subtype to provide a valid supertype instance in embedding applications.
+  - This may be especially difficult if a property type is defined as being an array of `oneOf` a given set of property type objects. Detecting which sub-schema applies to a given array element when accounting for dropped properties, **may** be a very difficult task.
+- Type extension might prove to have strange side-effects when considering link types which extend from one another.
+  - Say `Person` defines `Knows` and `Has Friend` links. If `Has Friend` extends `Knows`, then technically any `Has Friend` instance is also a `Knows` instance, and would need to satisfy the link constraints set on both. This has implications for `minItems` and `maxItems`, as well as for link destinations. If `Person` says it only `Knows` other `Person` entities, but `Has Friend` links to both `Person` and `Pet` entities, then a `Has Friend` link to a `Pet` entity would break the constraint set for `Knows`.
```

Comment:
I'm unsure if this is a **problem** yet.

It could be quite similar to the situations outlined above where the constraints are just combined, and in some cases that leads to an unsatisfiable condition (whereby the user then needs to update their type to make it satisfiable).

Reply:
Right. I think that's the appropriate resolution strategy. If people's definitions break logical transitivity principles we can prompt them to update those assumptions (updating types themselves, or forking/changing types).

Reply:
I was about to comment on this exact line prior to seeing your comment.
My instinct is that this is **not** a problem as the worst thing which could happen is an unsatisfiable constraint, which we explicitly opted out of.

```suggestion
- Type extension might prove to have strange side-effects when considering link types which extend from one another.
```

Reply:
Addressed in https://github.com/blockprotocol/blockprotocol/pull/1298